.github/workflows/self-update.yml

```diff
@@ -0,0 +1,38 @@
+name: self-update
+
+permissions: {}
+
+on:
+  workflow_dispatch:
+  schedule:
+    - cron: "0 0 * * *"
+
+jobs:
+  update:
+    runs-on: ubuntu-latest
+    steps:
+      - run: |
+          set -exuo pipefail
+
+          if [[ -z "$GH_TOKEN" ]]; then
+            echo "::error::GH_SELF_UPDATE_TOKEN secret is not set. Please either set this secret or disable this workflow."
+            exit 1
+          fi
+
+          gh auth setup-git
+          gh repo clone "$REPOSITORY" .
+
+          if ! git rev-parse remotes/upstream/HEAD &> /dev/null; then
+            echo "::error::This is not a fork. Please disable this workflow."
+            exit 1
+          fi
+
+          git config user.name "github-actions[bot]"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
```

Comment:
Nit: I'd be tempted to group this together with `gh auth setup-git`, since IMO configuring git's auth and git's `user` are related tasks.

E.g.
```sh
git config user.name "github-actions[bot]"
git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
gh auth setup-git

gh repo clone "$REPOSITORY" .
if ! git rev-parse remotes/upstream/HEAD &> /dev/null; then
```